src/saveload/ser.rs

```diff
@@ -1,166 +1,176 @@
+        serializer: S,
+    ) -> Result<S::Ok, S::Error>
+    where
+        E: Display,
+        S: Serializer,
+    {
+        let mut serseq = serializer.serialize_seq(None)?;
+        let ids = |entity| -> Option<M> { markers.get(entity).cloned() };
+        for (entity, marker) in (&*entities, &*markers).join() {
+            serseq.serialize_element(&EntityData::<M, Self::Data> {
+                marker: marker.clone(),
+                components: self.serialize_entity(entity, &ids)
+                    .map_err(ser::Error::custom)?, // TODO: revise
+            })?;
+        }
+        serseq.end()
     }
-    serseq.end()
-}
 
-/// This type implements `Serialize` so that it may be used in generic environment
-/// where `Serialize` implementation is expected.
-/// It may be constructed manually with `WorldSerialize::new`.
-/// Or fetched from `System` as `SystemData`.
-/// Serializes components in tuple `T` with marker `M`.
-#[derive(SystemData)]
-pub struct WorldSerialize<'a, M: Marker, E, T: Components<M::Identifier, E>> {
-    entities: Entities<'a>,
-    storages: <T as Storages<'a>>::ReadStorages,
-    markers: ReadStorage<'a, M>,
-    pd: PhantomData<E>,
+    /// Serialize components from specified storages via `SerializableComponent::save`
+    /// of all marked entities with provided serializer.
+    /// When the component gets serialized with `SerializableComponent::save` method
+    /// the closure passed in `ids` argument marks unmarked `Entity` (the marker of which was requested)
+    /// and it will get serialized recursively.
+    /// For serializing without such recursion see `serialize` function.
+    fn serialize_recursive<MS, S>(
+        &self,
+        entities: &EntitiesRes,
+        markers: &mut WriteStorage<M>,
+        allocator: &mut M::Allocator,
+        serializer: S,
+    ) -> Result<S::Ok, S::Error>
+    where
+        E: Display,
+        M: Marker,
+        S: Serializer,
+    {
+        let mut serseq = serializer.serialize_seq(None)?;
+        let mut to_serialize: Vec<(Entity, M)> = (entities, &*markers)
+            .join()
+            .map(|(e, m)| (e, m.clone()))
+            .collect();
+        while !to_serialize.is_empty() {
+            let mut add = vec![];
+            {
+                let mut ids = |entity| -> Option<M> {
+                    let (marker, added) = allocator.mark(entity, markers);
+                    if added {
+                        add.push((entity, marker.clone()));
+                    }
+                    Some(marker.clone())
+                };
+                for (entity, marker) in to_serialize {
+                    serseq.serialize_element(&EntityData::<M, Self::Data> {
+                        marker,
+                        components: self.serialize_entity(entity, &mut ids)
+                            .map_err(ser::Error::custom)?,
+                    })?;
+                }
+            }
+            to_serialize = add;
```

Comment:
This code is bit hard to decipher, but I don't know how to improve it.
Also you could reuse `to_serialize` and only do two allocations, but I don't know if it's worth it.